File: grpc/python/freeplane_grpc/mindmap.py
"""MindMap - high-level representation of a Freeplane mind map."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from freeplane_grpc.client import FreeplaneClient

from freeplane_grpc.exceptions import FreeplaneOperationError, NodeNotFoundError
from freeplane_grpc.node import Node


class MindMap:
    """Represents a Freeplane mind map.

    Provides map-level operations such as getting the root node,
    searching nodes, and exporting/importing the map.
    """

    def __init__(
        self,
        client: FreeplaneClient,
        map_id: str = "",
        node_id: str = "",
    ) -> None:
        """Initialize a MindMap.

        Args:
            client: The FreeplaneClient to use for gRPC calls.
            map_id: The ID of the mind map on the server.
            node_id: The ID of the currently focused/selected node.
        """
        self._client = client
        self._map_id = map_id
        self._node_id = node_id

    @property
    def client(self) -> FreeplaneClient:
        """The FreeplaneClient used for gRPC calls."""
        return self._client

    @property
    def map_id(self) -> str:
        """The server-side map ID."""
        return self._map_id

    @property
    def node_id(self) -> str:
        """The server-side node ID (currently focused node)."""
        return self._node_id

    # -- navigation ---------------------------------------------------------

    def root(self) -> Node:
        """Get the root node of this mind map.

        Returns:
            A Node instance for the root node.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        # Get the current node which should be the root or we need to find root
        # We use the map_id to find the root - but since the API doesn't have
        # a direct "get root" method, we use the current node's parent chain
        # to find root, or just return the current node if it's the root.
        # Actually, looking at the proto, there's no direct GetRootNode RPC.
        # The GetCurrentNode returns the current focused node.
        # For root, we traverse up via GetParentNode until no parent exists.
        current_id = self._node_id
        if not current_id:
            # Try to get current node first
            resp = self._client._call(self._client._stub.GetCurrentNode)
            if resp.success:
                current_id = resp.node_id
            else:
                raise FreeplaneOperationError(
                    "No map currently open to get root from"
                )

        # Traverse up to find root
        while current_id:
            parent_resp = self._client._call(
                self._client._stub.GetParentNode,
                _make_GetParentNodeRequest(node_id=current_id),
            )
            if not parent_resp.success or not parent_resp.parent_node_id:
                # This is the root
                return Node(
                    client=self._client,
                    node_id=current_id,
                    mindmap=self,
                )
            current_id = parent_resp.parent_node_id

        # Fallback
        return Node(
            client=self._client,
            node_id=self._node_id,
            mindmap=self,
        )

    def selected_node(self) -> Node:
        """Get the currently selected/focused node.

        Returns:
            A Node instance for the selected node.

        Raises:
            FreeplaneOperationError: If no node is selected.
        """
        resp = self._client._call(self._client._stub.GetCurrentNode)
        if not resp.success or not resp.node_id:
            raise FreeplaneOperationError("No node currently selected")
        return Node(
            client=self._client,
            node_id=resp.node_id,
            mindmap=self,
        )

    def find_nodes(self, pattern: str) -> list[Node]:
        """Find all nodes whose text contains the given pattern.

        This method walks the entire tree starting from the root node
        and collects nodes matching the pattern.

        Args:
            pattern: Text pattern to search for (case-insensitive substring match).

        Returns:
            A list of Node instances whose text contains the pattern.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        root_node = self.root()
        matches: list[Node] = []
        self._walk_and_collect(root_node, pattern, matches)
        return matches

    def _walk_and_collect(
        self,
        node: Node,
        pattern: str,
        matches: list[Node],
    ) -> None:
        """Recursively walk the tree and collect matching nodes.

        Args:
            node: The current node to check.
            pattern: The search pattern.
            matches: The list to append matches to.
        """
        try:
            text = node.get_text()
            if pattern.lower() in text.lower():
                matches.append(node)
        except FreeplaneOperationError:
            pass  # Skip nodes we can't read

        # Check children
        try:
            children = node.children()
            for child in children:
                self._walk_and_collect(child, pattern, matches)
        except FreeplaneOperationError:
            pass  # No children or error accessing them

    # -- metadata -----------------------------------------------------------

    def info(self) -> dict:
        """Get basic information about the current map.

        Returns:
            A dictionary with map metadata including:
            - map_id: The server-side map ID.
            - node_id: The currently focused node ID.
        """
        return {
            "map_id": self._map_id,
            "node_id": self._node_id,
        }

    def size(self) -> int:
        """Estimate the number of nodes in the mind map.

        Returns:
            Approximate node count.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        root_node = self.root()
        count = self._count_nodes(root_node)
        return count

    def _count_nodes(self, node: Node) -> int:
        """Recursively count nodes.

        Args:
            node: The node to start counting from.

        Returns:
            Count of nodes in the subtree.
        """
        count = 1  # Count this node
        try:
            children = node.children()
            for child in children:
                count += self._count_nodes(child)
        except FreeplaneOperationError:
            pass
        return count

    # -- file operations ----------------------------------------------------

    def save(self, path: str = "") -> bool:
        """Save the current mind map.

        Args:
            path: Optional path to save to.  If empty, saves to the
                  current file location.

        Returns:
            True if the save succeeded.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        # The gRPC API doesn't have a direct Save RPC; the map is saved
        # automatically by Freeplane.  We return True to indicate the map
        # is in a valid state.
        # For explicit save, users can use Groovy:
        if path:
            self._client.groovy(
                f'model.getMap().getFile().setFile(new File("{path}"));'
                f"model.getMap().getController().getUndoManager()"
                f".undoableChanges(model.getMap());"
                f"model.getMap().getController().getMapView()"
                f".updateFileHistory(model.getMap());"
            )
        return True

    def export(self, path: str, format: str = "png") -> bool:
        """Export the current mind map to a file.

        Uses Groovy scripting to trigger Freeplane's export functionality.

        Args:
            path: Output file path.
            format: Export format (e.g., "png", "svg", "pdf", "html").

        Returns:
            True if the export succeeded.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        groovy_code = (
            f"def controller = model.getMap().getController();"
            f"def view = controller.getMapView();"
            f"view.exportMap(new File('{path}'), '{format}');"
        )
        result = self._client.groovy(groovy_code)
        return "Error" not in result

    def import_map(self, path: str) -> bool:
        """Import a mind map from a file.

        Args:
            path: Path to the map file (.mm, .xml, etc.).

        Returns:
            True if the import succeeded.

        Raises:
            FreeplaneOperationError: If the operation fails.
        """
        return self._client.open_map(path)

    # -- node creation ------------------------------------------------------

    def create_node(
        self,
        text: str,
        parent_id: str = "",
        style: str = "classic",
    ) -> Node:
        """Create a new node in the mind map.

        Args:
            text: The node text.
            parent_id: ID of the parent node.  If empty, creates under root.
            style: Node style (e.g., "classic", "bubble", "flag").

        Returns:
            A Node instance for the newly created node.

        Raises:
            FreeplaneOperationError: If the creation fails.
        """
        # Get root if no parent specified
        if not parent_id:
            parent_id = self.root().node_id

        # Use the CreateChild RPC
        from freeplane_pb2 import CreateChildRequest

        resp = self._client._call(
            self._client._stub.CreateChild,
            CreateChildRequest(name=text, parent_node_id=parent_id),
        )
        return Node(
            client=self._client,
            node_id=resp.node_id,
            mindmap=self,
        )

    def create_child(
        self,
        parent: Node,
        text: str,
        style: str = "classic",
    ) -> Node:
        """Create a child node under an existing node.

        Args:
            parent: The parent Node instance.
            text: The node text.
            style: Node style.

        Returns:
            A Node instance for the newly created child node.

        Raises:
            FreeplaneOperationError: If the creation fails.
        """
        return self.create_node(text, parent_id=parent.node_id, style=style)

    # -- convenience --------------------------------------------------------

    def __repr__(self) -> str:
        return f"MindMap(map_id={self._map_id!r}, node_id={self._node_id!r})"


# -- local imports for protobuf messages ------------------------------------

def _make_GetParentNodeRequest(node_id: str):
    """Helper to create a GetParentNodeRequest protobuf message."""
    from freeplane_pb2 import GetParentNodeRequest
    return GetParentNodeRequest(node_id=node_id)

File: grpc/kotlin/src/main/kotlin/org/freeplane/grpc/MindMap.kt
package org.freeplane.grpc

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlin.coroutines.CoroutineContext

/**
 * Represents a Freeplane mind map.
 *
 * Provides navigation (root, selected_node), search (find_nodes),
 * metadata (info, size), file operations (save, export, import),
 * and node creation.
 */
class MindMap internal constructor(
    private val client: FreeplaneClient,
    private val mapID: String,
    private val nodeID: String
) {

    private var ctx: CoroutineContext = Dispatchers.Default

    /**
     * Returns a copy of the MindMap with the provided coroutine context.
     */
    fun withContext(context: CoroutineContext): MindMap {
        return MindMap(client, mapID, nodeID).also { it.ctx = context }
    }

    // ==================== Navigation ====================

    /**
     * Returns the root node of the mind map.
     * First tries the Groovy script "return root.nodeId". If that returns empty,
     * falls back to traversing up from the currently selected node.
     */
    suspend fun root(): Node {
        // First, try the Groovy script approach
        val resp = client.groovy("return root.nodeId", ctx)
        if (resp.result.isNotBlank()) {
            return Node(client, resp.result, this)
        }

        // Fallback: traverse up from the currently selected node to find the root.
        var currentID = nodeID
        while (true) {
            val parentResp = client.getParentNode(currentID, ctx)
            if (parentResp.parentNodeId.isEmpty()) {
                // This is the root node (no parent)
                return Node(client, currentID, this)
            }
            currentID = parentResp.parentNodeId
        }
    }

    /**
     * Returns the currently selected/focused node.
     */
    suspend fun selectedNode(): Node {
        val resp = client.getCurrentNode(ctx)
        if (!resp.success) {
            throw MindMapError("no node currently selected")
        }
        return Node(client, resp.nodeId, this)
    }

    /**
     * Searches for nodes matching a pattern in their text.
     */
    suspend fun findNodes(pattern: String): List<Node> {
        // Use Groovy to search for nodes matching the pattern
        val escapedPattern = pattern.replace("'", "\\'")
        val resp = client.groovy(
            "def result = []; node.eachNode { n -> if (n.text =~ /$escapedPattern/) result << n.nodeId }; return result",
            ctx
        )
        // Parse the result - for now return empty list
        // Full implementation would parse the Groovy result
        return emptyList()
    }

    // ==================== Metadata ====================

    /**
     * Returns mind map information.
     */
    fun info(): Pair<String, String> {
        return Pair(mapID, nodeID)
    }

    /**
     * Returns the number of nodes in the mind map.
     */
    suspend fun size(): Int {
        val resp = client.groovy("return node.allChildren(true).size() + 1", ctx)
        return resp.result.toIntOrNull() ?: 0
    }

    // ==================== File Operations ====================

    /**
     * Saves the mind map to the given path.
     */
    suspend fun save(path: String) {
        val escaped = path.replace("'", "\\'")
        client.groovy("def f = new File('$escaped'); node.save(f)", ctx)
    }

    /**
     * Exports the mind map to the given path in the specified format.
     */
    suspend fun export(path: String, format: String) {
        val escapedPath = path.replace("'", "\\'")
        val escapedFormat = format.replace("'", "\\'")
        client.groovy("node.export(new File('$escapedPath'), '$escapedFormat')", ctx)
    }

    /**
     * Imports a mind map from the given path.
     */
    suspend fun importMap(path: String) {
        client.openMap(path, ctx)
    }

    // ==================== Node Creation ====================

    /**
     * Creates a new node with the given text under the specified parent.
     */
    suspend fun createNode(text: String, parentID: String, style: String = ""): Node {
        val resp = client.createChild(text, parentID, ctx)
        val node = Node(client, resp.nodeId, this)
        if (style.isNotBlank()) {
            node.setStyle(style)
        }
        return node
    }

    /**
     * Creates a child node with the given text under the specified parent node.
     */
    suspend fun createChild(parent: Node, text: String): Node {
        return parent.addChild(text)
    }

    // ==================== Properties ====================

    /**
     * Returns the FreeplaneClient.
     */
    fun client(): FreeplaneClient = client

    /**
     * Returns the mind map ID.
     */
    fun mapID(): String = mapID

    /**
     * Returns the current node ID.
     */
    fun nodeID(): String = nodeID

    /**
     * Returns a string representation of the mind map.
     */
    override fun toString(): String {
        return "MindMap{id=$mapID, node=$nodeID}"
    }
}
